src/pkcs1v15.rs

```diff
@@ -299,24 +302,33 @@ impl SigningKey {
         self.hash
     }
 
-    pub fn new(key: RsaPrivateKey) -> Self {
+    pub(crate) fn digest(&self) -> Box<dyn DynDigest> {
+        self.digest.box_clone()
+    }
+
+    pub fn new(key: RsaPrivateKey, digest: Box<dyn DynDigest>) -> Self {
         Self {
             inner: key,
             hash: None,
+            digest,
         }
     }
 
-    pub fn new_with_hash(key: RsaPrivateKey, hash: Hash) -> Self {
+    pub fn new_with_hash(key: RsaPrivateKey, hash: Hash, digest: Box<dyn DynDigest>) -> Self {
```

Comment:
This seems problematic for a number of reasons.

You now have `hash` and `digest` parameters. "Hash" and "digest" are synonyms and in the examples you're passing effectively the same thing, it's just an enum variant versus a `Box`-ed digest instance. This isn't DRY.

I think you should pick one of `hash` or `digest`. With `hash`, you can pick the concrete `Digest` to use.

Why are you using `Box`? It erases the digest type, reducing type safety. If you choose to make everything use `digest`, I'd suggest introducing a generic parameter. That would be a more invasive change, however.

Reply:
I thought for using the generic parameter quite a while, while working on Signer/Verifier support. And in the end I decided against it. I wanted to have just two key types for all RSA signing keys (the PKCS1 v1.5 and PSS ones). In fact making the `SigningKey`/`VerifyingKey` accept the message rather than pre-hashed value just begs for that from my point of view. This way the all PKCS1 (and all PSS keys) parsed from the SPKI are type-compatible and can be stored and handled in a uniform way. One has to handle two major cases: PKCS1v1.5 vs PSS.

An alterntive of course would be to parse the SPKI.parameters in the app and to perform all the matching inside the app, handling the keys in the generic way in separate arms.

BTW: as we are talking about it, should I also add the Digest argument to the RSA Signature traits?
BTW2: I generally feel that rsa::Hash is misplaced or misimplemented. It either should go to pkcs1 crate or be reimplemented as a trait, which is implemented for different digest types. Or, maybe even better, I can split the Pkcs1v15 signer into padded and unpadded structs/traits/implementations.

Reply:
At the very least, there shouldn't be *both* `hash` and `digest` parameters.

Reply:
> BTW: as we are talking about it, should I also add the Digest argument to the RSA Signature traits?

Which traits are you referring to?

Are you talking about adding another `Box<Digest>`?

Reply:
@tarcieri nah, about adding `pss::Signature<D: Digest>` to supplement `pss::SignerKey<D: Digest>`.

Reply:
Oh, you're talking about the signature `struct`?

That's something I've considered but have not done yet with the `ecdsa` crate. The problem with parameterizing by a `D: Digest` (as opposed to an `enum` like `Hash`) is it complicates interoperability with things like hardware accelerators which may provide their own implementations of specific algorithms.

It's a tricky problem.

Reply:
On a quick note: I'm working on refactoring some JWT-related stuff in my code, and found that having the hash as a generic parameter helps, especially with allowing to implement `From<RsaPublicKey> for rsa::pas::VerifyingKey<H>`, etc.
I ended up building a wrapper in the meantime: https://github.com/matrix-org/matrix-authentication-service/blob/b6e92fb8f7ef81a61dcdee28d12e710743c88860/crates/jose/src/jwa/rsa.rs

Reply:
@sandhose I've pushed the #179, which implements the generified version of these structs + an RFC for further improvement. I'm going to close this PR.